### DISCUSSION TITLE
Question About my Pi5 Setup for Create3

### DISCUSSION THREAD

**Discussion by slowrunner** [iRobotEducation/create3_docs#544] [Ideas]
@alain-lnpcrd asked about my Pi5 configuration for running with Create3 https://github.com/iRobotEducation/create3_docs/discussions/472#discussioncomment-8452132 

>I have the impression that you have automated a lot of things for the RPi5 under bookworm (but maybe I'm just imagining things). So your directory, do you install it under your home folder or in a docker container? Do you then run a shell program that will install everything properly? ...
This is just to start correctly.

Like I told you there - I **FIRMLY** do not believe anyone should try to follow my path.  If you have a specific question about something you think might be useful, I'll consider it.

![WaLi_Architecture](https://github.com/iRobotEducation/create3_docs/assets/8950202/660993cb-b5c9-4b09-bb7e-542eb61ebcb9)

My robot controller is indeed running the Raspberry Pi OS Bookworm Desktop (which does not have ROS support),  
and follows a basic Raspberry Pi / Linux pattern of using crontab and systemd to start things:

Crontab at boot starts my "life logger"
```
@reboot sleep 10  && /usr/bin/python3 /home/pi/wali_pi5/plib/cleanlifelog.py
@reboot sleep 15  && /home/pi/wali_pi5/utils/nohup_loglife.sh
0 * * * * /usr/bin/python3 /home/pi/wali_pi5/plib/cleanlifelog.py
```

And systemd starts services:
- ip_feedback.service: IP address spoken by text to speech engine
```
[Unit]
Description=IP Feedback on Boot
After=syslog.target network.target


[Service]
Type=simple
user=pi
WorkingDirectory=/home/pi/
ExecStart=/bin/bash /home/pi/wali_pi5/utils/ip_feedback.sh

[Install]
WantedBy=multi-user.target
```

and starts:

- docker.r2hdp.service:  "ROS 2 Humble Desktop Plus" Docker container

```
[Unit]
Description=ROS 2 Humble Desktop Plus For WaLi Service
After=syslog.target network.target


[Service]
Type=simple
user=pi
WorkingDirectory=/home/pi/wali_pi5/c3ws
# Stop r2hdp container (if running)
ExecStartPre=-/usr/bin/docker stop r2hdp
# Remove the r2hdp container (if exists)
ExecStartPre=-/usr/bin/docker rm r2hdp
# Wait to allow life logging to start up at boot time
ExecStartPre=/bin/sleep 60
# Start r2hdp detached (which runs start_wali.sh)
ExecStart=/bin/bash /home/pi/wali_pi5/configs/docker/start_docker_detached_r2hdp.sh

[Install]
WantedBy=multi-user.target
```

The start_docker_detached_r2hdp.sh script actually starts Docker:

```
#!/bin/bash

cd /home/pi/wali_pi5/c3ws
# --rm removes container after exit
# --w working dir to start in
echo -e "\n*** STARTING ROS 2 HUMBLE PLUS IN DOCKER ***"
# docker run -dit --name r2hdp --net=host --device /dev/snd \
docker run -dit --name r2hdp --net=host \
 -v /dev/snd:/dev/snd \
 -v /home/pi:/home/pi -v /dev/input:/dev/input \
 -v /dev/bus/usb:/dev/bus/usb  \
 -e TZ=America/New_York \
 -w /home/pi/wali_pi5/c3ws --privileged r2hdp 
# -v /dev/bus/usb:/dev/bus/usb  -w /home/pi/wali_pi5/c3ws --privileged --rm r2hdp &

```

which  runs the start_wali.sh script:

```
#!/bin/bash

# This script is run when Docker container r2hdp starts
echo -e "START_WALI.SH EXECUTING"
cd /home/pi/wali_pi5/c3ws
source install/setup.bash

echo -e "\n*** Start F710 game controller node"
echo '*** ros2 launch teleop_twist_joy teleop-launch.py joy_config:="F710" '
ros2 launch teleop_twist_joy teleop-launch.py joy_config:="F710" &

echo -e "\n*** Start odometer node"
echo '*** ros2 run wali odometer & '
ros2 run wali odometer &

echo -e "\n*** Start wali.say_node"
echo '*** ros2 run wali say_node &'
ros2 run wali say_node &

echo -e "\n **** STARTING ir2scan "
echo -e "executing: ros2 run wali ir2scan &"
ros2 run wali ir2scan &

# MUST START IN FOREGROUND or docker will exit!
ros2 run wali wali_node

# Use ./attach_to_docker_r2hdp.sh to connect to this terminal
# Use .term_to_r2hdp.sh to open a new terminal session 
#    in the ROS 2 Humble Desktop Plus container
```

NOTE: THIS IS NOT GOOD ROS PRACTICE - Wali should be a lifecycle node with a launch file, but I don't know how to do that, so for the time being I just run most nodes in the background, and run the wali_node in foreground so the Docker container won't quit.

I am not going to help you install "Wali" on your Pi5 - ain't going to happen.  

You mentioned you are over 70, feeling pressure of advancing age, and the concept of elder care robots.  I have a lot of thoughts on these topics, but since this forum is about the Create3 - projects and problems, I will simply tell you that I am also over 70 and  have no delusions that I am capable of writing any software for the Create3 robot that will perform any function useful to me (or my wife).  (The only slightly useful function of my other 24/7 robot, Carl, is responding to "Hey Carl, weather?" which I certainly did not NEED to build a robot to accomplish.

**Comment by alain-lnpcrd** [comment]
Thanks @slowrunner
There is a lot of information that interests me in what you send me.
Currently I'm going through the turtlebot4 documentation and I'm encountering a great bug (perhaps linked to my humble configuration) in the examples given in turtlebot4-python-tutorials (turtlebot4/turtlebot4-tutorials) relating to the functioning of the feedback of the actions launched. Basically, the feedback does not accompany the action, or even the feedback ends before the action starts.
But I am unable to understand and fix the bug. I am still very far from having the knowledge.
